Implementation/Services/index.md

```diff
@@ -0,0 +1,510 @@
+using System.Linq;
+using System.Web;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco8.Services
+{
+        public interface ISiteHelperService
+    {
+        IPublishedContent GetNewsSection();
+        IPublishedContent GetContactUsPage();
+    }
+}
+```
+Create the concrete service class that implements the interface:
+```csharp
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+using Umbraco.Core.Models.PublishedContent;
+
+namespace Umbraco8.Services
+{
+    public class SiteHelperService : ISiteHelperService
+    {
+        public IPublishedContent GetNewsSection()
+        {
+            //ToDo: implement this!
+            throw new NotImplementedException();
+        }
+        public IPublishedContent GetContactUsPage()
+        {
+            //ToDo: implement this!
+            throw new NotImplementedException();
+        }
+    }
+}
+```
+
+Register the custom service with Umbraco's underlying DI container using an IUserComposer:
+
+```csharp
+using Umbraco.Core;
+using Umbraco.Core.Composing;
+using Umbraco8.Services;
+
+namespace Umbraco8.Composers
+{
+    public class RegisterSiteHelperServiceComposer : IUserComposer
+    {
+        public void Compose(Composition composition)
+        {         
+            composition.Register<ISiteHelperService, SiteHelperService>(Lifetime.Request);          
+        }
+    }
+}
+```
+
+Note we register the service in 'Lifetime.Request' scope - as we'll be making use of the transient UmbracoContext
+
+Now, implementing the service, we can inject Umbraco's core helper IPublishedContentQuery (the same as you would use when writing Umbraco.Content(1234)) into our custom service, to use in order to locate the special site pages we are after:
+
+```csharp
+using System.Linq;
+using Umbraco.Core.Models.PublishedContent;
+using Umbraco.Web;
+
+namespace Umbraco8.Services
+{
+    public class SiteHelperService : ISiteHelperService
+    {
+        private readonly IPublishedContentQuery _contentQuery;
+        public SiteHelperService(IPublishedContentQuery contentQuery)
+        {
+            _contentQuery = contentQuery;
+        }
+        public IPublishedContent GetNewsSection()
+        {
+            var siteRoot = _contentQuery.ContentAtRoot().FirstOrDefault();
+            var newsSection = siteRoot.Children(f => f.ContentType.Alias == "newsSection").FirstOrDefault();
+            return newsSection;
+        }
+        public IPublishedContent GetContactUsPage()
+        {
+            var siteRoot = _contentQuery.ContentAtRoot().FirstOrDefault();
+            var contactUs = siteRoot.Children(f => f.ContentType.Alias == "contactUs").FirstOrDefault();
+            return contactUs;
+        }
+    }
+}
+```
+
+:::Note
+If we planned to use our SiteHelperService outside of a request where UmbracoContext is not available we'd use the UmbracoContextFactory + EnsureUmbracoContext() methods described above to access the content cache.
```

Comment:
Based on my previous note, it is preferable to not use Request based services. What happens now if this current service is needed on a background thread or you want to be able to inject this into a singleton - well you can't!

Instead, registering things as singletons where possible makes the most sense. Then if you want your object to have access to the UmbracoContext, you will inject the `IUmbracoContextAccessor` ... in fact, this is best practices. We don't want people to inject UmbracoContext or IPublishedContentQuery directly - the only places this is 'ok' is in Controllers - but as noted already, these things are already injected into controllers.

For custom services, best practice is register as a singleton, if you need an UmbracoContext, inject `IUmbracoContextAccessor`. Then it's also important to check if there is one available. For example, `if (umbCtxAccessor.UmbracoContext == null) { //there is no current context, so now you should use the UmbracoContextFactory }`. However with that said, you could simply always just use the `UmbracoContextFactory` instead since this already checks for the existence of a current umb context on the `IUmbracoContextAccessor`. The benefit however of using `IUmbracoContextAccessor` is that you know if you are within an http request or not.

Reply:
Ok on the first point, the subtle thing is we 'do' want people to create custom services that are only used in Views/Controllers! - that is the inspiration for the article - Yes people have access to UmbracoHelper in a controller or view - but it leads to implementation of logic in the view or controller - eg Umbraco.ContentAtRoot().FirstOrDefault().Children(f=>f.IsVisible() && f.ContentType.Alias == "landingPage" && f.Children.Any());  which is sort of fine when the implementation logic is written once, in one place, but you tend to find duplication and subtle variations on implementation building up overtime, bring a lack of consistency, where somebody forgot 'where else' they pulled back news articles in date order - but what can be a neater alternative, is if this logic is put into a Service to be called in multiple places - it's the first step towards some degree of separation. WIth V8 the wonderful thing is the complicated bit of wiring up DI is done for peoples - so I'm trying to highlight in a Murky rather than super Clean way, how easy it is to get those benefits of separation, without it getting super complicated. The familiarity of the IPublishedContentQuery methods is helpful!

But agree the warnings against using the custom service outside of a http request aren't clear...
... and agree it's just as easy to use the IPublishedContentCache methods ... it's just the unfamiliarity - I wonder if over time we could sync the method naming a little?

... anyway my suggested solution for now is to include two examples, one showing injecting the familiar interface methods and then add the warning, and then show the 'better practice' way, again my hope is here, people are lured in by the easier example first and then conclude themselves that the second example is much better, and not so hard to use to avoid getting caught out.

I feel if we only show the second example, then the first example will be used by people anyway and they'll run into trouble, but this might not be the approach the official documentation should take to engaging people on this subject.

With IUmbracoContextAccessor, I haven't added anything - feel like we should, but wasn't clear of the benefit/how to explain best, eg if there is no UmbracoContext you'd have to use the UmbracoContextFactory approach, so you'd have to inject both into your custom service, unless you were sure you were only going to use your service where there is an UmbracoContext... then you could just use UmbracoContextAccessor - but if the UmbracoContextFactory always checks via the UmbracoContextAccessor anyway - it might be simplest to say just use that always? but I sense I'm failing to grasp the significance of knowing whether you are in a http request or not inside the context of your custom service :-(

 (perhaps there could be a flag on the UmbracoContextReference object that indicates whether you are or not?)